[Feature][Frontend]: Add support for `stream_options` in `ChatCompletionRequest`

Etelis opened the pull request:
This PR introduces support for the `stream_options` parameter in the `ChatCompletionRequest` class, addressing the feature request in issue #4967.

### Changes
- Added the `StreamOptions` class with an optional `include_usage` attribute.
- Modified the `ChatCompletionRequest` class to include the `stream_options` attribute.
- Updated the `chat_completion_stream_generator` method to handle the `stream_options` parameter, allowing optional inclusion of usage statistics in streaming responses.

Fix #4967 (https://github.com/vllm-project/vllm/issues/4967)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR needs to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code needs to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project stays correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM users understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feels confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status updates every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code>action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

Etelis committed:
Add stream_options in ChatCompletionRequest

Add StreamOptions Class

Add stream_options validation in ChatCompletionRequest

Etelis committed:
[Feature]: Support `stream_options` option (#4967)

- Introduced the `StreamOptions` class in `OpenAIBaseModel` with an optional `include_usage` attribute.

- Added `stream_options` attribute to the `ChatCompletionRequest` class, defaulting to `None`.

Etelis committed:
-Implemented a validator to ensure `stream_options` can only be set if `stream` is true.

Etelis committed:
[Feature]: Add optional usage statistics to streaming responses

- Updated `chat_completion_stream_generator` to include support for `stream_options` with an `include_usage` flag.
- Modified the initial response generation to conditionally include `usage` field based on `stream_options.include_usage`.
- Enhanced the token-by-token and finish responses to conditionally include `usage` field if `stream_options.include_usage` is set.
- Added a final usage statistics message if `stream_options.include_usage` is set, including prompt tokens and completion tokens.

DarkLight1337 commented:
The code looks good at first glance, but to verify this, we should add some unit tests? In particular, test these cases for both Chat Completion and Completion APIs:

- `stream=True, stream_options=None`: No usage info returned
- `stream=True, stream_options={"include_usage": False}`: No usage info returned
- `stream=True, stream_options={"include_usage": True}`: Usage info returned; check its contents
- `stream=False, stream_options={"include_usage": None}`: Bad request/validation error
- ~`stream=False, stream_options=None`: Usage info returned; check its contents~ Already covered by existing tests

Etelis commented:
> The code looks good at first glance, but to verify this, let's add some unit tests:
> 
>     * `stream=True, stream_options=None`: No usage info returned
> 
>     * `stream=True, stream_options={"include_usage": False}`: No usage info returned
> 
>     * `stream=True, stream_options={"include_usage": True}`: Usage info returned; check its contents
> 
>     * `stream=False, stream_options={"include_usage": None}`: Bad request/validation error
> 
>     * ~`stream=False, stream_options=None`: Usage info returned; check its contents~ Already covered by existing tests
> 
> 
> Could you also update the Completions API handler with `stream_options` support? I think the logic should pretty much be the same.

I will try to implement the tests today. 
Do you mind me splitting the Completions API to another PR? for clarity..

DarkLight1337 commented:
> Do you mind me splitting the Completions API to another PR? for clarity.

Feel free to do so!

Etelis committed:
Tests added for the following scenarios:

- stream=True, stream_options=None

- stream=True, stream_options={"include_usage": True}

- stream=True, stream_options={"include_usage": False}

- stream=False, stream_options={"include_usage": None}

- stream=False, stream_options={"include_usage": False}

- stream=False, stream_options={"include_usage": True}

Etelis commented:
Added the unit tests.
@DarkLight1337

DarkLight1337 commented:
> Added the unit tests. @DarkLight1337

It seems that there is something wrong with the new schema definition, based on the failing tests.

Etelis committed:
Forgot to notate validator as a classmethod.

Etelis committed:
Merge remote-tracking branch 'remote/main' into feature/support-stream-options#4967

Etelis committed:
Fixed failing tests based on:
- Duplicated Usage defenition in protocol.py.
- Line too long in several files.

Etelis committed:
Fixed failing tests:
- Yapf formating.

Etelis committed:
FIxing failing tests:
- yapf in protocol file.

DarkLight1337 reviewed (approved):
Looks good now, thanks for the hard work!
